wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt

```diff
@@ -448,3 +486,39 @@ private class SourceCodeHelper(
     return m
   }
 }
+
+private class SourceInfo(
+  val helper: SourceCodeHelper,
+  path: List<Int> = emptyList(),
+) {
+  constructor(
+    fileDescriptor: FileDescriptorProto,
+    path: List<Int> = emptyList()
+  ) : this(SourceCodeHelper(fileDescriptor), path)
+
+  private val path = mutableListOf(*path.toTypedArray())
+
+  fun push(value: Int) {
+    path.add(value)
+  }
+
+  fun info(): LocationAndComments {
+    return helper.getLocation(path)
+  }
+
+  fun new(): SourceInfo {
+    return SourceInfo(helper, listOf(*path.toTypedArray()))
+  }
+}
+
+private fun <T> List<T>.withSourceInfo(sourceInfo: SourceInfo, value: Int): List<Pair<SourceInfo, T>> {
+  val baseSource = sourceInfo.new()
+  val result = mutableListOf<Pair<SourceInfo, T>>()
+  baseSource.push(value)
+  for ((index, elem) in withIndex()) {
+    val newSource = baseSource.new()
```

Comment:
Another approach, instead of explicit clone and then push, is to use the `+` operator for lists. It turns out that it's implemented even for immutable lists and thus always returns a copy. Take it or leave it. This way is fine, too. It was just something I learned when I was also looking at how to express the path construction stuff more cleanly.

Reply:
Let me fully enclose this stuff and we can optimize the classes functionality after